src/main/java/pl/agh/edu/master_diet/service/SearchService.java

```diff
@@ -0,0 +1,27 @@
+package pl.agh.edu.master_diet.service;
+
+import pl.agh.edu.master_diet.core.model.database.Browsable;
+import pl.agh.edu.master_diet.core.model.shared.SearchResult;
+import pl.agh.edu.master_diet.repository.BrowserRepository;
+
+import java.util.List;
+
+public abstract class SearchService<Searched extends Browsable> {
+
+    private final BrowserService<Searched> browserService;
+    private final PageFittingService<Searched> pageFittingService;
+
+
+    public SearchService(BrowserRepository<Searched> browserRepository, PageFittingService<Searched> fittingService) {
+        this.browserService = new BrowserService<>(browserRepository);
+        this.pageFittingService = fittingService;
+    }
+
+    public SearchResult<Searched> searchBrowsable(String searchTerm, Integer pageIndex, Integer perPage) {
+        List<Searched> searchedProducts = browserService.search(searchTerm);
+        Integer maximumPageNumber = pageFittingService.calculateMaximumPageNumber(searchedProducts, perPage);
+        searchedProducts = pageFittingService.adjustListToPageInfo(searchedProducts, pageIndex, perPage);
+        return new SearchResult<>(searchedProducts, maximumPageNumber);
```

Comment:
What I mentioned, to me usage of builder would be better

Reply:
Generic builder problem. Can you give me some examples of why builder usage is better? It's interesting.

Reply:
Already discussed, changed to builder